solutions/POI 05-sza.mdx

```diff
@@ -0,0 +1,218 @@
+---
+id: poi-05-sza
+source: POI 2005
+title: Template
+author: Andi Qu
+---
+
+Let the string be $S$.
+
+<Spoiler title="Hint 1">
+
+A template of $S$ is both a prefix and a suffix of $S$. This suggests that we should use KMP.
+
+</Spoiler>
+
+<Spoiler title="Hint 2">
+
+If $T$ is a template of $S$ and $U$ is a template of $T$, then $U$ is also a template of $S$.
+
+<Spoiler title="Proof">
+
+Since $U$ is a template of $T$, we can make $T$ using $U$. Since $T$ is a template of $S$, we can make $S$ using $T$.
+
+Therefore, we can also make $S$ using $U$ (i.e. it's a template of $S$).
+
+</Spoiler>
+
+</Spoiler>
+
+<Spoiler title="Hint 3">
+
+First, use KMP to compute the prefix function $\pi$ of $S$. The shortest template of $S$ is either $S$ itself or the shortest template of $S[0 : \pi[|S| - 1]]$.
+
+<Spoiler title="Proof">
+    
+We'll prove instead that any template of $S$ that isn't $S$ itself is also a template of $S[0 : \pi[|S| - 1]]$, since that's equivalent.
+
+Since $S[0 : \pi[|S| - 1]]$ is the longest proper prefix of $S$ that is also a suffix, we know that the only longer string that can be $S$'s template is $S$ itself. Let the shortest template of $S[0 : \pi[|S| - 1]]$ be $T'$ and some longer template of $S$ be $T$.
+
+By hint 1, $T$ must be a prefix and suffix of $S$, which means that it's also a prefix and suffix of $S[0 : \pi[|S| - 1]]$.
+
+This means that we can make at least half of $T$'s prefix using $T'$, since we can make $S[0 : \pi[|S| - 1]]$ (which is bigger than $T$) using $T'$ and $|T'| < |T|$. A similar argument holds for being able to make at least half of $T$'s suffix.
+
+This implies that $T'$ is a template of $T$, as we can combine this prefix and suffix to form the whole string. Therefore $T$ is also a template of $S[0 : \pi[|S| - 1]]$.
```

Comment:
> This implies that $T'$ is a template of $T$, as we can combine this prefix and suffix to form the whole string. 

$T'$ is a template of $T$, therefore $T'$ is a template of $S$. But that's not what you're trying to show?

What if $|T'|>|T|$? (You assumed this, but you need to show why this is impossible.)